src/tree/criterion.ts

```diff
@@ -0,0 +1,128 @@
+import { assert } from '../typesUtils'
+import { int } from '../randUtils'
+
+export type ImpurityMeasure = 'gini' | 'entropy'
+export interface SampleData {
+  sample_number: int
+  current_feature_value: number
+}
+
+export function GiniCoefficient(label_freqs: int[], n_samples_: int) {
+  let freq_squares = 0
+  for (let i = 0; i < label_freqs.length; i++) {
+    freq_squares += label_freqs[i] * label_freqs[i]
+  }
+  return 1 - freq_squares / (n_samples_ * n_samples_)
+}
+
+export function Entropy(label_freqs: int[], n_samples: int) {
+  let entropy = 0
+  for (let i = 0; i < label_freqs.length; i++) {
+    let label_frequency = label_freqs[i]
+    if (label_frequency > 0) {
+      label_frequency /= n_samples
+      entropy -= label_frequency * Math.log2(label_frequency)
+    }
+  }
+  return entropy
+}
+
+export class Criterion {
+  label_data_: int[]
+  impurity_measure_: ImpurityMeasure
+  impurity_fn_: (label_freqs: int[], n_samples: int) => number
+  start_: int = 0
+  end_: int = 0
+  pos_: int = 0
+  n_labels_: int
+  label_freqs_total_: int[] = []
+  label_freqs_left_: int[] = []
+  label_freqs_right_: int[] = []
+  n_samples_: int = 0
+  n_samples_left_: int = 0
+  n_samples_right_: int = 0
+
+  constructor(impurity_measure: ImpurityMeasure, label_data: number[]) {
+    assert(
+      ['gini', 'entropy'].includes(impurity_measure),
+      'Unkown impurity measure. Only supports gini, and entropy'
+    )
+
+    this.impurity_measure_ = impurity_measure
+    if (this.impurity_measure_ === 'gini') {
+      this.impurity_fn_ = GiniCoefficient
+    } else {
+      this.impurity_fn_ = Entropy
+    }
+    // This assumes that the labels are 0,1,2,...,(n-1)
+    this.n_labels_ = Math.max(...label_data) + 1
```

Comment:
Super concise but it may cause a stack-overflow, if `label_data` becomes too large (~1_000_000 elements but it may be less if the stack is full).

Reply:
Agreed, I'll make the change here.